### ISSUE TITLE
Too many warning = crash

### ISSUE THREAD

**Issue by frog42** [AtomLinter/linter-stylint#74]
Somewhere around 64 warnings in a document causes this to crash.

**Comment by ViZhe** [comment]
Need more information. The problem is originated only in stylint?

![errors](https://cloud.githubusercontent.com/assets/8091902/20619936/76c8eb42-b310-11e6-91ec-9a5e35cc2a8b.png)

**Comment by frog42** [comment]
Weird. Here's the error I get:

Error parsing stylint output
Something went wrong parsing the stylint output, please see the console for details.

Then in the console, there are a lot of warnings and it starts with this:

init.js [sm]:123 linter-stylint: Failed to parse output as JSON [{"filePath":"style.styl","messages":[{"severity":"Warning","message":"hexidecimal color should be a variable","rule":"colors","line":38,"column":8}, [then a bunch of similar warnings]

If I fix enough warnings, it works again.

**Comment by ViZhe** [comment]
Really weird. At first glance the JSON looks good.

Can you give an example of breaking the stylus file and stylintrc?

**Comment by frog42** [comment]
I'm linking a gist that breaks it for me. If you comment out ".imageBox" around line 700, it doesn't give me an error when saving. And if I fix enough warnings, I can have the whole thing work without it giving me the error.

https://gist.github.com/frog42/8c4b6d4371813a8aa324034e868e4dc3

**Comment by ViZhe** [comment]
I uncomment all the lines, but it succeeded, at least with my stylintrc and in strict mode.

The atom and the packages are the latest versions? OS?

Throw off, please, your stylintrc and the full JSON from the log.

**Comment by frog42** [comment]
Fresh install of OSX Sierra. Fresh install of Atom. No settings changed from default in linter-stylint. Stylint installed using "npm install styling -g". Not using a stylintrc config file, just the default.

Maybe found the problem. The very bottom of the console error message ends like this: 

{"severity":"Warning","message":"prefer alphabetical when sorting properties","rule":"sortOrder","line":712},{"severity":

It's cutting off the JSON output. 

Here's the start of of the error message:

linter-stylint: /Users/frog42/.atom/packages/linter-stylint/lib/init.js:132

**Comment by frog42** [comment]
Oh, and latest version of Atom (1.12.5) was installed using Homebrew: brew cask install atom

**Comment by frog42** [comment]
Ah, and running stylint from command line on the file works fine.

**Comment by ViZhe** [comment]
Technically, this package gets a json from [atom-linter](https://github.com/AtomLinter/linter-stylint/blob/master/lib/init.js#L6). You can try to replace [execNode](https://github.com/AtomLinter/linter-stylint/blob/master/lib/init.js#L101) to `exec` in `/.atom/packages/linter-stylint/lib/init.js`. Don't forget to restart the Atom.

Node version?

If this not help then we try to call steelbrain.

**Comment by frog42** [comment]
No difference if I change execNode to exec. I'm running 7.2.0 of Node.

I just did a fresh Sierra install using only the following commands and it still gives the same error:

ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)”
brew doctor
brew update
brew tap homebrew/dupes
brew tap caskroom/cask
brew tap caskroom/versions
brew cask install atom

curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.32.1/install.sh | bash
nvm install node
nvm use node
npm install npm -g
npm install pug -g
npm install coffee-script -g
npm install stylus -g
npm install nib -g
npm install jeet -g
npm install rupture -g
npm install axis -g

apm install stylus
apm install linter
apm install linter-stylint


Running stylint from command line works fine.

**Comment by ViZhe** [comment]
Hey, @steelbrain

Here a strange problem with mac os, execNode (exec too) returns the [trimmed json](https://github.com/AtomLinter/linter-stylint/issues/74#issuecomment-263071806). I tried to repeat this on my mint and windows, but all in vain.

Have you encountered similar behavior?

**Comment by steelbrain** [comment]
Yeah we trim it on purpose, does trimming introduce any problems?

**Comment by ViZhe** [comment]
probably I not so expressed ;(
no trimmed, **cutted**.

as I understand it, 
if the number of errors is less than ~60, then @frog42 [get from execNode](https://github.com/AtomLinter/linter-stylint/blob/master/lib/init.js#L101) valid JSON.
If more than ~60, then here is:

```
[{
  "filePath":"style.styl",
  "messages":[
    // correct the object of warning
    {"severity":"Warning","message":"hexidecimal color should be a variable","rule":"colors","line":38,"column":8},
    // then ~60 same correct objects
    {...},
    // then json is cut off
    {"severity":

```

**Comment by Arcanemagus** [comment]
We've seen this bug before in a few other linters, if I remember correctly this is due to them calling `system.exit(code)` directly, instead of simply setting the exit code and letting the program end. The issue this causes is the `stdout` buffer of the process (`stylint` in this case) doesn't get fully flushed out, so it ends up getting cut off. The reason you don't see this on the CLI is that your terminal is probably "fixing" this for you and continuing to write the buffer after the process died.

@ViZhe The difference between `exec` and `execNode` is that the first one expects the program to be able to run on its own, while the second is expecting a Node.js script (as it passes it as a parameter to the Atom executable running in "Node.js mode"). Changing the code to `exec()` would break things on Windows, but would transparently work mostly the same on *NIX.

**Comment by SimenB** [comment]
We do call `process.exit` internally, which does seem unnecessary (although node@0.10 doesn't support `process.exitCode`). I'll try to get it fixed in V2, although in V2 you can use the programmatic API instead of invoking the CLI, if that's better

**Comment by jamb84** [comment]
We found the issue comes down to an 8192 byte limit on responses returned to stdout through child_process.spawn. Either needs to be swapped to a non-limited processor or the stylint maintainers should implement chunking in the output. For now if you modify the file "linter-stylint/node_modules/stylint/src/core/done.js" to chunk the output:

msg += this.cache.msg

                if ( msg ) {
                        var fs = require('fs');
                        var chunks = msg.match(/.{1,10}/g);
                        for (var i = 0; i < chunks.length; i++) {
                                fs.writeSync(process.stdout.fd, chunks[i]);
                        }
                }

this should resolve your issue until someone else fixes the problem officially .

**Comment by dhniels** [comment]
I'm having this issue as well. My _variables.styl is full of linter errors saying `variables and parameters must be prefixed with the $ sign (prefixVarsWithDollar)`

anytime I save a stylus file I get the error:
```
Error parsing stylint output
Something went wrong parsing the stylint output, please see the console for details.
```
